assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/MetricsOverrideBinding.java

```diff
@@ -0,0 +1,118 @@
+ * CheMajorWebSocketEndpointExecutorServiceProvider} and {@link
+ * CheMinorWebSocketEndpointExecutorServiceProvider}.
+ *
+ * <p>{@link ExecutorService} that is returned by these providers will publish metrics to {@link
+ * PrometheusMeterRegistry}
+ */
+public class MetricsOverrideBinding implements Module {
+  @Override
+  public void configure(Binder binder) {
+    binder
+        .bind(CheMajorWebSocketEndpointExecutorServiceProvider.class)
+        .to(MeteredCheMajorWebSocketEndpointExecutorServiceProvider.class);
+
+    binder
+        .bind(CheMinorWebSocketEndpointExecutorServiceProvider.class)
+        .to(MeteredCheMinorWebSocketEndpointExecutorServiceProvider.class);
+  }
+
+  @Singleton
+  public static class MeteredCheMajorWebSocketEndpointExecutorServiceProvider
+      extends CheMajorWebSocketEndpointExecutorServiceProvider {
+
+    private final PrometheusMeterRegistry meterRegistry;
+
+    private ExecutorService executorService;
+
+    @Inject
+    public MeteredCheMajorWebSocketEndpointExecutorServiceProvider(
+        @Named("che.core.jsonrpc.processor_core_pool_size") int corePoolSize,
+        @Named("che.core.jsonrpc.processor_max_pool_size") int maxPoolSize,
+        @Named("che.core.jsonrpc.processor_queue_capacity") int queueCapacity,
+        PrometheusMeterRegistry meterRegistry) {
+      super(corePoolSize, maxPoolSize, queueCapacity);
+      this.meterRegistry = meterRegistry;
+    }
+
+    @Override
+    public ExecutorService get() {
+      if (executorService == null) {
+        synchronized (MeteredCheMajorWebSocketEndpointExecutorServiceProvider.class) {
+          if (executorService == null) {
+            executorService =
+                ExecutorServiceMetrics.monitor(
+                    meterRegistry,
+                    super.get(),
+                    CheMajorWebSocketEndpointConfiguration.EXECUTOR_NAME,
+                    Tags.empty());
+          }
+        }
+      }
+      return executorService;
+    }
+  }
+
+  @Singleton
+  public static class MeteredCheMinorWebSocketEndpointExecutorServiceProvider
+      extends CheMinorWebSocketEndpointExecutorServiceProvider {
+    private final PrometheusMeterRegistry meterRegistry;
+
+    private ExecutorService executorService;
+
+    @Inject
+    public MeteredCheMinorWebSocketEndpointExecutorServiceProvider(
+        @Named("che.core.jsonrpc.minor_processor_core_pool_size") int corePoolSize,
+        @Named("che.core.jsonrpc.minor_processor_max_pool_size") int maxPoolSize,
+        @Named("che.core.jsonrpc.minor_processor_queue_capacity") int queueCapacity,
+        PrometheusMeterRegistry meterRegistry) {
+      super(corePoolSize, maxPoolSize, queueCapacity);
+      this.meterRegistry = meterRegistry;
+    }
+
+    @Override
+    public ExecutorService get() {
+      if (executorService == null) {
+        synchronized (MeteredCheMinorWebSocketEndpointExecutorServiceProvider.class) {
```

Comment:
Not that it is likely, but generally speaking synchronizing on objects accessible outside of the class is dangerous because it can cause deadlocks in unsuspecting code outside of control of this class.

Consider:
```java
synchronized(MeteredCheMinorWebSocketEndpointExecutorServiceProvider.class) {
  Thread t = new Thread(() -> new MeteredCheMinorWebSocketEndpointExecutorServiceProvider().get());
  t.start();
  t.join();
}
```

Reply:
I would rather make ```public synchronized ExecutorService get``` Otherwise it might be a small overengineering of this problem.
@sleshchenko @metlos agree?